src/client/Microsoft.Identity.Client/Instance/Region/RegionManager.cs

```diff
@@ -149,94 +152,130 @@ private bool IsTelemetryRecorded(ApiEvent apiEvent)
+                            logger.Verbose($"[Region discovery] {s_regionDiscoveryDetails}");
+                        }
+                    }
+                }
+
+                result = result ?? new RegionInfo(null, RegionAutodetectionSource.FailedAutoDiscovery, s_regionDiscoveryDetails);
+            }
+            finally
+            {
+                s_failedAutoDiscovery = result.RegionSource == RegionAutodetectionSource.FailedAutoDiscovery;
+                s_autoDiscoveredRegion = result.Region;
+                s_regionDiscoveryDetails = result.RegionDetails;
+
+                _lockDiscover.Release();
             }
 
-            try
-            {
-                var headers = new Dictionary<string, string>
-                {
-                    { "Metadata", "true" }
-                };
-
-                Uri imdsUri = BuildImdsUri(DefaultApiVersion);
-
-                HttpResponse response = await _httpManager.SendGetAsync(imdsUri, headers, logger, retry: false, GetCancellationToken(requestCancellationToken))
-                    .ConfigureAwait(false);
-
-                // A bad request occurs when the version in the IMDS call is no longer supported.
-                if (response.StatusCode == HttpStatusCode.BadRequest)
-                {
-                    string apiVersion = await GetImdsUriApiVersionAsync(logger, headers, requestCancellationToken).ConfigureAwait(false); // Get the latest version
-                    imdsUri = BuildImdsUri(apiVersion);
-                    response = await _httpManager.SendGetAsync(BuildImdsUri(apiVersion), headers, logger, retry: false, GetCancellationToken(requestCancellationToken))
-                        .ConfigureAwait(false); // Call again with updated version
-                }
-
-                if (response.StatusCode == HttpStatusCode.OK && !response.Body.IsNullOrEmpty())
-                {
-                    region = response.Body;
-
-                    if (ValidateRegion(region, $"IMDS call to {imdsUri.AbsoluteUri}", logger))
-                    {
-                        logger.Info($"[Region discovery] Call to local IMDS succeeded. Region: {region}. {DateTime.UtcNow}");
-                        return new RegionInfo(region, RegionAutodetectionSource.Imds, null);
-                    }
-                }
-                else
-                {
-                    s_regionDiscoveryDetails = $"Call to local IMDS failed with status code {response.StatusCode} or an empty response. {DateTime.UtcNow}";
-                    logger.Verbose($"[Region discovery] {s_regionDiscoveryDetails}");
-                }
+            return result;
+        }
 
-            }
-            catch (Exception e)
+        // returns cached region if any.
+        // if nothing is cached, returns null.
+        private RegionInfo GetCachedRegion(ICoreLogger logger)
+        {
+            if (s_failedAutoDiscovery)
```

Comment:
Does it make sense to add a test around this, when auto discovery has already failed it is not performed again?

Reply:
It is a private method. So as long as corresponding public method is checked, it is fine.